File: banchus-app/src/main/java/pe/nanamochi/banchus/service/infra/IPApiClient.java
package pe.nanamochi.banchus.service.infra;

import java.net.InetAddress;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import pe.nanamochi.banchus.domain.dto.Geolocation;

@Slf4j
@Service
@RequiredArgsConstructor
public class IPApiClient {
  private final RestTemplate restTemplate;

  public Geolocation fetchFromIP(InetAddress ip) {
    try {
      String url = "https://ip-api.com/json/" + ip.getHostAddress();
      Geolocation geo = restTemplate.getForObject(url, Geolocation.class);

      if (geo != null && "success".equals(geo.status())) {
        return geo;
      }

      log.warn("IP-API returned non-success status for IP: {}", ip.getHostAddress());
    } catch (Exception e) {
      log.error("Failed to fetch geolocation for IP: {}", ip.getHostAddress(), e);
    }

    return Geolocation.local();
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/infrastructure/clients/IPApiClient.kt
package pe.nanamochi.banchus.infrastructure.clients

import java.net.InetAddress
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import org.springframework.web.client.getForObject
import pe.nanamochi.banchus.dto.external.Geolocation

@Service
class IPApiClient(private val restTemplate: RestTemplate) {
    private val log = LoggerFactory.getLogger(javaClass)

    fun fetchFromIP(ip: InetAddress): Geolocation {
        val hostAddress = ip.hostAddress
        val url = "http://ip-api.com/json/$hostAddress"

        return runCatching { restTemplate.getForObject<Geolocation>(url) }
            .getOrNull()
            ?.takeIf { it.status == "success" }
            ?: run {
                log.warn(
                    "IP-API failed or returned non-success for IP: $hostAddress - defaulting to local geolocation."
                )
                Geolocation.local()
            }
    }
}
